controller/iteration.go

```diff
@@ -232,6 +233,92 @@ func (c *IterationController) Update(ctx *app.UpdateIterationContext) error {
+		itr, err := appl.Iterations().Load(ctx.Context, ctx.IterationID)
+		if err != nil {
+			return jsonapi.JSONErrorResponse(ctx, err)
+		}
+		s, err := appl.Spaces().Load(ctx, itr.SpaceID)
+		if err != nil {
+			return jsonapi.JSONErrorResponse(ctx, goa.ErrNotFound(err.Error()))
+		}
+		if !uuid.Equal(*currentUser, s.OwnerId) {
+			log.Warn(ctx, map[string]interface{}{
+				"space_id":     s.ID,
+				"space_owner":  s.OwnerId,
+				"current_user": *currentUser,
+			}, "user is not the space owner")
+			return jsonapi.JSONErrorResponse(ctx, errors.NewForbiddenError("user is not the space owner"))
+		}
+		if c.IsRoot(itr, s) {
+			log.Warn(ctx, map[string]interface{}{
+				"space_id":     s.ID,
+				"iteration_id": itr.ID,
+			}, "can not delete root iteration")
+			return jsonapi.JSONErrorResponse(ctx, errors.NewForbiddenError("can not delete root iteration"))
+		}
+		subtree, err := appl.Iterations().LoadChildren(ctx, ctx.IterationID)
+		if err != nil {
+			return jsonapi.JSONErrorResponse(ctx, goa.ErrInternal(err.Error()))
+		}
+		// delete all children along with given iteration
+		subtree = append(subtree, *itr)
+		for _, child := range subtree {
+			err = appl.Iterations().Delete(ctx.Context, child.ID)
+			if err != nil {
+				log.Error(ctx, map[string]interface{}{
+					"iteration_id": child.ID,
+					"err":          err.Error(),
+				}, "unable to delete iteration")
+				return jsonapi.JSONErrorResponse(ctx, errors.NewInternalError(ctx, err))
+			}
+			// fetch associated work items
+			wis, err := appl.WorkItems().LoadByIteration(ctx, child.ID)
+			if err != nil {
+				return jsonapi.JSONErrorResponse(ctx, errors.NewInternalError(ctx, err))
+			}
+			// set root iteration to associated work items
+			for _, wi := range wis {
+				// to be safe, load root iteration of this WI and then assign it
+				ri, err := appl.Iterations().Root(ctx, wi.SpaceID)
+				if err != nil {
+					log.Error(ctx, map[string]interface{}{
+						"space_id": wi.SpaceID,
+						"err":      err.Error(),
+					}, "unable to get root iteration for space")
+					return jsonapi.JSONErrorResponse(ctx, errors.NewInternalError(ctx, err))
+				}
+				// move WI to root iteration
+				wi.Fields[workitem.SystemIteration] = ri.ID.String()
+				_, err = appl.WorkItems().Save(ctx, wi.SpaceID, *wi, *currentUser)
+				if err != nil {
+					log.Error(ctx, map[string]interface{}{
+						"workitem_id": wi.ID,
+						"err":         err.Error(),
+					}, "unable to update iteration for work item")
+					return jsonapi.JSONErrorResponse(ctx, errors.NewInternalError(ctx, err))
+				}
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return jsonapi.JSONErrorResponse(ctx, errors.NewInternalError(ctx, err))
+	}
+	return ctx.OK([]byte{})
+}
+
+// IsRoot Checks if given iteration is a root iteration or not
+func (c *IterationController) IsRoot(itr *iteration.Iteration, sp *space.Space) bool {
+	return (itr.SpaceID == sp.ID && itr.Path.String() == path.SepInService)
```

Comment:
why do you need to verify `itr.SpaceID == sp.ID` ?

Reply:
Right, it must not be checked but I feel safe doing that and feel it more readable.
And it resembels the query `tx := m.db.Where("space_id = ? and path = ?", spaceID, "").First(&itr)` to GET root iteration of that space.
Should I remove the check of space ID?
please see https://github.com/fabric8-services/fabric8-wit/pull/1683/commits/21a922cabd1ddca24311e954ec69455efd52b824